### PR TITLE
feat(BB): Initialization of lane constructor code

### PR DESCRIPTION
**Pull request by davidterpay**
*THIS PR WILL NOT PASS CI UNTIL THE OTHER PRS ARE MERGED INTO HERE.*

## Overview 
The lane constructor is a generic struct that developers should utilize to create new lanes. It wraps all of the required interface functionality defined in `lane_interface.go` into a generic type of lane. What this means is that users no longer have to worry about re-implementing a lot of the basic things needed to build a lane. 

Instead of implementing the entire set of interfaces, they only have to implement a set of a few handlers (if any) and a single tx priority object.

### MatchHandler
This lets the constructor know which lanes to match transactions to. You implement that and pass it into the lane and now it will always accept those transactions.

### PrepareLaneHandler
This allows the lane to reap transactions from its mempool, have arbitrary inclusion functionality, and returns the txs you want to include in the proposal, the txs you need to remove from the lane, and an error if something is wrong. You'll see that the `LaneConstructor` utilizes this and then updates the proposal as well as that lane's mempool. This allows developers to not have to worry about mempool bloat and imo simplies the amount of responsibility on the developer in general. 

### ProcessLaneHandler
This does something similar to the above except it just ingresses transactions and validates them according to the logic of the lane. It must always return the remaining transactions after it is done processing the ones that belong to it. These transactions will then be passed onto the next lane.

### CheckOrderHandler
This validates that transactions that belong to the lane and are included in proposals are ordered correctly. The constructor lane also wraps this functionality

A good amount of this PR is just moving things around and creating the final file structure that makes sense. 

### LaneConstructorMempool
This just ingresses a `TxPriority` object that lets the mempool do all of the logic it needs with as minimal amount of info as possible. I found a small bug in the priority nonce mempool implementation that the SDK has so I am going to keep it here for the time being.

### CODE REVIEW

File: blockbuster/abci/abci.go [hunk 1]
```diff
@@ -125,9 +131,14 @@ func ChainPrepareLanes(chain ...blockbuster.Lane) blockbuster.PrepareLanesHandle
 		// Cache the context in the case where any of the lanes fail to prepare the proposal.
 		cacheCtx, write := ctx.CacheContext()
 
+		// We utilize a recover to handler any panics or errors that occur during the preparation
```
Comment (nivasan1) [inline blockbuster/abci/abci.go:134]: handle

File: blockbuster/lane_abci.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	if err := utils.RemoveTxsFromLane(txsToRemove, l); err != nil {
+		l.Logger().Error(
+			"failed to remove transactions from lane",
+			"lane", l.Name(),
+			"err", err,
+		)
+	}
+
+	// Update the proposal with the selected transactions.
+	if err := proposal.UpdateProposal(l, txs); err != nil {
```
Comment (nivasan1) [inline blockbuster/lane_abci.go:33]: Feel like the PrepareLane methods shld only return txs (and the top level Chainer would handle constructing the proposal). I don't like how the proposal / insertion logic is passed into each lane's PrepareLane

File: blockbuster/priority_nonce.go [hunk 3]
```diff
@@ -313,14 +313,16 @@ func (i *PriorityNonceIterator[C]) Next() sdkmempool.Iterator {
...
-		return i.iteratePriority()
-	} else if i.mempool.cfg.TxPriority.Compare(key.priority, i.nextPriority) == 0 {
-		// Weight is incorporated into the priority index key only (not sender index)
-		// so we must fetch it here from the scores map.
-		weight := i.mempool.scores[txMeta[C]{nonce: key.nonce, sender: key.sender}].weight
-		if i.mempool.cfg.TxPriority.Compare(weight, i.priorityNode.Next().Key().(txMeta[C]).weight) < 0 {
+	if i.priorityNode.Next() != nil {
+		if i.mempool.cfg.TxPriority.Compare(key.priority, i.nextPriority) < 0 {
 			return i.iteratePriority()
+		} else if i.mempool.cfg.TxPriority.Compare(key.priority, i.nextPriority) == 0 {
```
Comment (nivasan1) [inline blockbuster/priority_nonce.go:319]: kek... this was from upstream, no?

**Review by nivasan1 — APPROVED** [review]
lgtm

**Comment by nivasan1** [comment]
nit: I don't think the LaneConstructor name is appropriate, it isn't a constructor... it's really a `BaseLane` implementation no?

File: blockbuster/priority_nonce.go (hunk 3 above)
Comment (davidterpay) [reply blockbuster/priority_nonce.go:319]: yea it was

File: blockbuster/lane_abci.go (hunk 2 above)
Comment (davidterpay) [reply blockbuster/lane_abci.go:33]: resolved offline

**Pull request merged**